### PR TITLE
Stabilize Sample analysis API

### PR DESCRIPTION
**Pull request by IgnatBeresnev**
Part of #3112, depends on #3184.

Moved the sample analysis API from internal to public, wrote documentation and a bunch of tests (obvious ones, corner cases + documentation contracts).

### CODE REVIEW

File: subprojects/analysis-kotlin-descriptors/compiler/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/descriptors/compiler/CompilerDescriptorAnalysisPlugin.kt [hunk 1]
```diff
@@ -75,13 +75,8 @@ public class CompilerDescriptorAnalysisPlugin : DokkaPlugin() {
         plugin<InternalKotlinAnalysisPlugin>().fullClassHierarchyBuilder providing { DescriptorFullClassHierarchyBuilder() }
     }
 
-    /**
-     * StdLib has its own a sample provider
-     * So it should have a possibility to override this extension
-     */
-    @InternalDokkaApi
-    public val kotlinSampleProviderFactory: Extension<SampleProviderFactory, *, *> by extending {
```
Comment (IgnatBeresnev) [inline subprojects/analysis-kotlin-descriptors/compiler/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/descriptors/compiler/CompilerDescriptorAnalysisPlugin.kt:83]: Need to discuss what to do with stdlib

File: subprojects/analysis-kotlin-descriptors/compiler/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/descriptors/compiler/impl/DescriptorSampleAnalysisEnvironment.kt
Comment (IgnatBeresnev) [inline subprojects/analysis-kotlin-descriptors/compiler/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/descriptors/compiler/impl/DescriptorSampleAnalysisEnvironment.kt:1]: This file isn't actually new, but I renamed it from `KotlinSampleProvider.kt`, and git / idea thought it was a new one :\

File: subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/services/SymbolSampleAnalysisEnvironment.kt
Comment (IgnatBeresnev) [inline subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/services/SymbolSampleAnalysisEnvironment.kt:1]: This file also isn't new, I also renamed it from `KotlinSampleProvider.kt`, although I did introduce some refactoring changes to understand what the code does better.

File: subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/services/SymbolSampleAnalysisEnvironment.kt [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+    private fun findPsiElement(sourceSet: DokkaSourceSet, fqLink: String): PsiElement? {
+        val analysisContext = kotlinAnalysis[sourceSet]
+        return analyze(analysisContext.mainModule) {
+            val isRootPackage = !fqLink.contains('.')
+            // supposed because we expect the last element to be a top-level function, not class/class+function
+            val supposedFunctionName = if (isRootPackage) fqLink else fqLink.substringAfterLast(".")
+            val supposedPackageName = if (isRootPackage) "" else fqLink.substringBeforeLast(".")
+
+            getTopLevelCallableSymbols(FqName(supposedPackageName), Name.identifier(supposedFunctionName)).firstOrNull()?.psi
```
Comment (IgnatBeresnev) [inline subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/services/SymbolSampleAnalysisEnvironment.kt:87]: The called function `getTopLevelCallableSymbols` returns top-level functions only, which is not as useful as what we have in the Descriptors implementation.

Is there a way to get any symbol here by fully qualified name? That is, a Kotlin or a Java class or anything else?

I've added some user-friendly warning log messages that only Kotlin functions are allowed, or that links to Java declarations are not allowed, but I didn't figure out how to do the same with symbols API.

Any ideas / pointers?

**Comment by IgnatBeresnev** [comment]
Squashed commits that fixed the API dump and some tests, but no major changes.

**Review by vmishenev — COMMENTED** [review]
It should support stdlib.

File: subprojects/analysis-kotlin-api/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/sample/SampleSnippet.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2014-2023 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package org.jetbrains.dokka.analysis.kotlin.sample
+
+/**
+ * Represents a sample code snippet of a Kotlin function. The snippet includes both file
+ * import statements and the sample function body.
```
[the remark this answers is not in the document]
Comment (IgnatBeresnev) [reply subprojects/analysis-kotlin-api/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/sample/SampleSnippet.kt:9]: Fixed the KDoc to address both comments.

I Believe the list returns all file imports, regardless of whether they are used or not - not even sure there's a way for us to know if they are used.

File: subprojects/analysis-kotlin-descriptors/compiler/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/descriptors/compiler/impl/DescriptorSampleAnalysisEnvironment.kt [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+import org.jetbrains.kotlin.resolve.source.KotlinSourceElement
+
+internal class DescriptorSampleAnalysisEnvironmentCreator(
+    private val context: DokkaContext,
+) : SampleAnalysisEnvironmentCreator {
+
+    private val descriptorAnalysisPlugin = context.plugin<CompilerDescriptorAnalysisPlugin>()
+
+    override fun <T> use(block: SampleAnalysisEnvironment.() -> T): T {
+        return runBlocking(Dispatchers.Default) {
```
[the remark this answers is not in the document]
Comment (IgnatBeresnev) [reply subprojects/analysis-kotlin-descriptors/compiler/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/descriptors/compiler/impl/DescriptorSampleAnalysisEnvironment.kt:44]: Huh. You might've been commenting on the old commit or something? I didn't change anything, but the comment is there

File: subprojects/analysis-kotlin-api/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/sample/SampleAnalysisEnvironment.kt [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+ * Fully-configured and ready-to-use sample analysis environment.
+ *
+ * It's best to limit the scope of use and lifetime of this environment as it takes up
+ * additional resources which could be freed once the samples have been analyzed.
+ * Therefore, it's best to use it through the [SampleAnalysisEnvironmentCreator.use] lambda.
+ */
+public interface SampleAnalysisEnvironment {
+
+    /**
+     * Resolves a Kotlin sample function by its fully qualified name, and returns its import statements and body.
```
[the remark this answers is not in the document]
Comment (IgnatBeresnev) [reply subprojects/analysis-kotlin-api/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/sample/SampleAnalysisEnvironment.kt:23]: > We should support the @snippet tag.

I don't think there's enough demand/pressure to support the `@snippet` tag in Dokka itself, let alone to provide stable analysis API for it :) It would be nice to have, but it should definitely be addressed separately (likely as part of #3005).

File: subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/services/SymbolSampleAnalysisEnvironment.kt [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+    override fun resolveSample(sourceSet: DokkaSourceSet, fullyQualifiedLink: String): SampleSnippet? {
+        val psiElement = findPsiElement(sourceSet, fullyQualifiedLink)
+        if (psiElement == null) {
+            dokkaLogger.warn(
+                "Unable to resolve a @sample link: \"$fullyQualifiedLink\". Is it used correctly? " +
+                        "Expecting a link to a reachable (resolvable) top-level Kotlin function."
+            )
+            return null
+        } else if (psiElement.containingFile !is KtFile) {
```
[the remark this answers is not in the document]
Comment (IgnatBeresnev) [reply subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/services/SymbolSampleAnalysisEnvironment.kt:65]: oof, tbh it looks quite scary. `KtSymbol` returns an enum, and judging by the comment for `KtSymbol#psi`, multiple entries can mean the symbol has sources:

> If [origin] is one of [KtSymbolOrigin.SOURCE], [KtSymbolOrigin.JAVA], [KtSymbolOrigin.LIBRARY] then the source is not null

So it looks like the check here should be 
```kotlin
if (ktSymbol.origin in listOf(KtSymbolOrigin.SOURCE, KtSymbolOrigin.JAVA, KtSymbolOrigin.LIBRARY))
```

to which my immediate question would be: what happens if another enum value is added without us noticing, and samples stop working in some corner case with a false positive?

> containingFile can have undefined behaviour, e.g. for synthesized declaration

Could you clarify what you mean by undefined behaviour? The Javadoc for it states

> @return the file instance, or null if the PSI element is not contained in a file

which sounds like what we need, does it not?

As a side-note while I still remember: we also rely on `psi.containingFile` in `processImports()`

File: subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/services/SymbolSampleAnalysisEnvironment.kt (hunk 2 above)
Comment (IgnatBeresnev) [reply subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/services/SymbolSampleAnalysisEnvironment.kt]: > There is resolveKDocTextLink.

Had to do a bit of refactoring (I needed to get `KtSymbol` and not `DRI`), but it works, thanks! Commit here: https://github.com/Kotlin/dokka/commit/92aa97ae4727d3d086ef2481f223eb493a8d1103

> But for Kotlin it should resolve only functions (not only top-level), see https://kotlinlang.org/docs/kotlin-doc.html#sample-identifier

Yep, there are test cases for it, and they all pass now, so everything should be correct. I needed to resolve any link (even if it's not a function) to provide a user-friendly warning message.

**Comment by IgnatBeresnev** [comment]
Rebased onto master, so I had to force push, but didn't squash any commits.

Also addressed the review comments, so it's ready for another round

File: subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/services/SymbolSampleAnalysisEnvironment.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (vmishenev) [reply subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/services/SymbolSampleAnalysisEnvironment.kt:66]: > if (ktSymbol.origin in listOf(KtSymbolOrigin.SOURCE, KtSymbolOrigin.JAVA, KtSymbolOrigin.LIBRARY))

We need only `KtSymbolOrigin.SOURCE` for now. 
 `For [KtSymbolOrigin.LIBRARY] the generated by Kotlin class file source element is returned`

> what happens if another enum value is added without us noticing, and samples stop working in some corner case with a false positive?

You can use `KtSymbol.sourcePsi()` or use exhaustive `when`.

> Could you clarify what you mean by undefined behaviour? The Javadoc for it states 

I mean `KtSymbol.psi`

```
For other [KtSymbolOrigin] behaviour is undefined
```

File: subprojects/analysis-kotlin-descriptors/compiler/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/descriptors/compiler/CompilerDescriptorAnalysisPlugin.kt (hunk 1 above)
Comment (IgnatBeresnev) [reply subprojects/analysis-kotlin-descriptors/compiler/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/descriptors/compiler/CompilerDescriptorAnalysisPlugin.kt]: Will be addressed in #3108

File: subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/services/SymbolSampleAnalysisEnvironment.kt (hunk 6 above)
[the remark this answers is not in the document]
Comment (IgnatBeresnev) [reply subprojects/analysis-kotlin-symbols/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/symbols/services/SymbolSampleAnalysisEnvironment.kt]: Changed it to `sourcePsiSafe()`: https://github.com/Kotlin/dokka/commit/79a411d41f9d171d5d0638721cdf1e906b9f00a6#diff-bfbfeaf57d9a071f5398bc3ab5cfaed15af09554a31d2b06ba3810d6bbb935f1R87

File: plugins/base/src/main/kotlin/transformers/pages/DefaultSamplesTransformer.kt [hunk 7]
```diff
@@ -51,14 +52,14 @@ internal class DefaultSamplesTransformer(val context: DokkaContext) : PageTransf
...
-        sample: SampleProvider.SampleSnippet,
+        sample: SampleSnippet,
     ): ContentNode {
         val node = contentCode(contentPage.content.sourceSets, contentPage.dri, createSampleBody(sample.imports, sample.body), "kotlin")
         return dfs(fqLink, node)
     }
 
-    fun createSampleBody(imports: String, body: String) =
-        """ |$imports
+    private fun createSampleBody(imports: List<String>, body: String) =
```
[the remark this answers is not in the document]
Comment (IgnatBeresnev) [reply plugins/base/src/main/kotlin/transformers/pages/DefaultSamplesTransformer.kt]: > why do we need .takeIf { it.isNotEmpty() } here? if list is empty, message from joinToString will be empty

Left a comment in https://github.com/Kotlin/dokka/commit/ea88a0a6e6666b31b4a4071f77c80889ac22d9af. TLDR is that if `imports` are empty, there should be no empty/blank lines. But if `imports` are present, there should be an empty line between them and the body.

It's very hacky, and I want to refactor this whole class, but this PR is not small as it is, so I tried to avoid unnecessary changes. Hopefully it'll be completely re-written in #3041 anyway

> Also, default separator in joinToString is , - it should be replaced with empty string or new line
> Looks like there is no tests for this :)

We have some tests (which is how I remember about this place and why I added the condition), but none of them have multiple import lines :) So you indeed found a bug. I've changed one of the tests to check it now, but again, hopefully better tests will be added as part of #3041

File: subprojects/analysis-kotlin-api/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/sample/SampleSnippet.kt [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+ * Copyright 2014-2023 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license.
+ */
+
+package org.jetbrains.dokka.analysis.kotlin.sample
+
+/**
+ * Represents a sample code snippet of a Kotlin function. The snippet includes both file
+ * import directives (all, even unused) and the sample function body.
+ *
+ * @property imports list of import statement values, without the `import` prefix.
```
[the remark this answers is not in the document]
Comment (IgnatBeresnev) [reply subprojects/analysis-kotlin-api/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/sample/SampleSnippet.kt]: On one hand, it's a good point, I didn't think about it. On the other, it's impossible to have an `internal` constructor (unless we make it an interface), and after giving it some thought - I don't think it makes sense to check such things, it's indeed up to the implementation.

We have `kotlin-analysis-api` tests, which should (and hopefully do) check for all of the corner cases described in these KDocs, and these tests are supposed to be an implementation-independent test suite like JCK. So anyone who implements the analysis api can take these tests and check if their implementation is correct (from our point of view).

We can check minor things like trimming in runtime, but we can't check if the `body` contains a function or even Kotlin code, as the implementation can return anything, so we'll still have to have tests, and I would prefer to have all checks be in one place.

And also someone might want to return an "invalid" object here, it's not uncommon to have very specialized implementations of the JVM that are not "by specification", but have some other huge benefits :)

File: plugins/base/src/main/kotlin/transformers/pages/DefaultSamplesTransformer.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (whyoleg) [reply plugins/base/src/main/kotlin/transformers/pages/DefaultSamplesTransformer.kt]: Thanks, now I see the difference. It will be good then to add comment to #3041 about additional tasks to follow-up

**Review by whyoleg — APPROVED** [review]

File: dokka-subprojects/analysis-kotlin-descriptors-compiler/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/descriptors/compiler/impl/DescriptorSampleAnalysisEnvironment.kt [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+        val kdocLink = kdocFinder.resolveKDocLink(
+            fromDescriptor = packageDescriptor,
+            qualifiedName = fqLink,
+            sourceSet = dokkaSourceSet,
+            emptyBindingContext = true
+        ).firstOrNull()
+
+        if (kdocLink == null) {
+            dokkaLogger.warn(
+                "Unable to resolve a @sample link: \"$fqLink\". Is it used correctly? " +
```
Comment (vmishenev) [inline dokka-subprojects/analysis-kotlin-descriptors-compiler/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/descriptors/compiler/impl/DescriptorSampleAnalysisEnvironment.kt:110]: I missed in the first review the message does not have the location of sample link. It can be useful to find the link in the input code.

File: plugins/base/src/main/kotlin/transformers/pages/DefaultSamplesTransformer.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (IgnatBeresnev) [reply plugins/base/src/main/kotlin/transformers/pages/DefaultSamplesTransformer.kt]: Added it to the issue's description that it's expected to have the refactoring and that the tests will need to be revised

File: subprojects/analysis-kotlin-descriptors/compiler/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/descriptors/compiler/impl/DescriptorSampleAnalysisEnvironment.kt
Comment (IgnatBeresnev) [reply subprojects/analysis-kotlin-descriptors/compiler/src/main/kotlin/org/jetbrains/dokka/analysis/kotlin/descriptors/compiler/impl/DescriptorSampleAnalysisEnvironment.kt:1]: Yeah, I just needed to move the file manually using `git mv` instead of drag&drop in IDEA. Will for sure pay attention to it in the future

In this particular case, I think it's OK though - after the review rounds, I ended up changing so much that it's as if the class was new
